src/Orleans/Lifecycle/ILifecycleObserver.cs

```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading.Tasks;
+
+namespace Orleans
+{
+    public interface ILifecycleObserver
+    {
+        Task OnStart();
+        Task OnStop();
```

Comment:
Not sure I agree that there should be an OnStart and OnStop (or setup/teardown), instead of having the lifecycle enum have all the stages itself, including the ones when it's tearing down, and exceptional stages. Still undecided what's best (at least for grain lifecycle)